testsuite/cluster-test/src/atomic_histogram.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::tx_emitter::TxStats;
+use std::{
+    fmt,
+    fmt::Debug,
+    ops::Sub,
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+const DEFAULT_HISTOGRAM_CAPACITY: usize = 2048;
+const DEFAULT_HISTOGRAM_STEP_MILLIS: u64 = 50;
+
+pub struct AtomicHistogramAccumulator {
+    capacity: usize,
+    step_millis: u64,
+    buckets: Vec<AtomicU64>,
+}
+
+impl Default for AtomicHistogramAccumulator {
+    fn default() -> AtomicHistogramAccumulator {
+        AtomicHistogramAccumulator::new(DEFAULT_HISTOGRAM_CAPACITY, DEFAULT_HISTOGRAM_STEP_MILLIS)
+    }
+}
+
+impl AtomicHistogramAccumulator {
+    pub fn new(size: usize, step: u64) -> AtomicHistogramAccumulator {
+        let mut buf = Vec::with_capacity(size);
+        for _i in 0..size {
+            buf.push(AtomicU64::new(0));
+        }
+        Self {
+            capacity: size,
+            step_millis: step,
+            buckets: buf,
+        }
+    }
+
+    pub fn snapshot(&self) -> AtomicHistogramSnapshot {
+        let mut buf = Vec::with_capacity(self.capacity);
+        for i in 0..self.capacity {
+            buf.push(self.buckets[i].load(Ordering::Relaxed));
+        }
+        AtomicHistogramSnapshot {
+            capacity: self.capacity,
+            step_millis: self.step_millis,
+            buckets: buf,
+        }
+    }
+
+    fn get_bucket_num(&self, latency: u64) -> u64 {
+        let bucket_num = latency / self.step_millis;
+        if bucket_num >= self.capacity as u64 - 2 {
+            return self.capacity as u64 - 1;
+        }
+        bucket_num
+    }
+
+    pub fn record_latency_ms(&self, latency_ms: u64, num_requests: u64) {
+        let bucket_num = self.get_bucket_num(latency_ms);
+        self.buckets[bucket_num as usize].fetch_add(num_requests as u64, Ordering::Relaxed);
+    }
+}
+
+pub struct AtomicHistogramSnapshot {
+    capacity: usize,
+    step_millis: u64,
+    pub(crate) buckets: Vec<u64>,
+}
+
+impl Default for AtomicHistogramSnapshot {
+    fn default() -> AtomicHistogramSnapshot {
+        AtomicHistogramAccumulator::new(DEFAULT_HISTOGRAM_CAPACITY, DEFAULT_HISTOGRAM_STEP_MILLIS)
+            .snapshot()
+    }
+}
+
+// reserving place for debug print in the future
+impl Debug for AtomicHistogramSnapshot {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "",)
+    }
+}
+
+pub fn sub_vec(a: &[u64], b: &[u64]) -> Vec<u64> {
```

Comment:
This should not be public function? In fact, maybe it does not need to be separate function at all, can just be code inside AtomicHistogramSnapshot::sub?

Additionally, I think you can just enforce that snapshots should be of same length, e.g. just write `assert!`(a.len==b.len, "Histogram snapshots must have same size, a:{}, b:{}", a.len, b.len)`

Reply:
we can not use assert here, because the first snapshot sub, we are using NONE to be the prev_snap

Reply:
Not sure I understand - what is None? Don't we just use AtomicHistogramSnapshot::default?

Reply:
in main.rs, we record prev_stats for duration average:
```
        let mut prev_stats: Option<TxStats> = None;
        while Instant::now() < deadline {
            let window = Duration::from_secs(10);
            tokio::time::delay_for(window).await;
            let stats = emitter.peek_job_stats(&job);
            let delta = &stats - &prev_stats.unwrap_or_default();
            prev_stats = Some(stats);
            println!("{}", delta.rate(window));
        }

so for the first duration of stat substract, we do expect the length is 0

Reply:
`unwrap_or_default` returns Default, and default is not a zero-size snapshot, it's snapshot of default size filled with zeroes....

Reply:
oh, I c, good to know ```unwrap_or_default``` return default

Reply:
Yep (this is also what name suggests :) )
You could not pass None in this function even if you wanted to - argument is `&Snapshot` - you can not pass None with this type (it has to be Option to allow for None)